### PR TITLE
implement ring2xy in pure python

### PR DESCRIPTION
**Pull request by nbren12**
this is equivalent to ring2hpd

### CODE REVIEW

File: padding_debug.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (nbren12) [inline padding_debug.py:1]: remove b4 merging. possibly make example showing the use of local2xy.

File: earth2grid/healpix/nn.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (nbren12) [inline earth2grid/healpix/nn.py:1]: refactored to avoid a circular import error.

**Comment by nbren12** [comment]
Here is an example showing the local padding procedure

image

File: examples/sphinx_gallery/local_padding.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    # Choose a point of interest
+    lon0, lat0 = 10, 85  # Example point in northern hemisphere
+
+    # Extract local patch
+    pix = extract_local_patch(nside, lon0, lat0, pad)
+
+    # Visualize results
+    plot_global_data(z)
+    plot_local_patch(lon, lat, z, pix, lon0, lat0)
```
Comment (nbren12) [inline examples/sphinx_gallery/local_padding.py:125]: image

File: earth2grid/healpix/_padding/pure_python.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        (3, 2) -> (0, 2)
+        """
+        f_west = torch.where(is_missing_s_pole_tile & (x1 >= y1), east_face, f1)
+        # x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), 1(x1-y1) %nside, x1)
+        x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), (x1 - y1 - 1) % nside, x1)
+        x_west = torch.where(is_missing_s_pole_tile & (x1 == y1), 0, x_west)
+        y_west = y1
+        return x_west, y_west, f_west
+
+    x_west, y_west, f_west = _pad_from_east(x_west, y_west, east_face, f_west)
```
Comment (akshaysubr) [inline earth2grid/healpix/_padding/pure_python.py:78]: Is this correct? (from_east vs from_west)

File: earth2grid/healpix/nn.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    Applies a 2D convolution over an input image composed of several input
+    planes.
+
+    This operator supports :ref:`TensorFloat32<tf32_on_ampere>`.
+
+    See :class:`~torch.nn.Conv2d` for details and output shape.
+
+    """
+    px, py = padding
```
Comment (akshaysubr) [inline earth2grid/healpix/nn.py:34]: padding is defined as just an int in the signature but used as a tuple?

File: earth2grid/healpix/nn.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    if px != py:
+        raise ValueError(f"Padding should be equal in x and y, got px={px}, py={py}")
+
+    n, c, x, y = input.shape
```
Comment (akshaysubr) [inline earth2grid/healpix/nn.py:38]: No f dim?

File: examples/sphinx_gallery/benchmark_healpix_pad_timing.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+nside = 128
+
+
+neval = 10
+
+
+def test_func(label, pad, compile=False):
+    # warm up
+    if compile:
+        pad = torch.compile(pad)
```
Comment (akshaysubr) [inline examples/sphinx_gallery/benchmark_healpix_pad_timing.py:46]: Is there any situation where someone might not want to run with torch.compile? If not, might be better to just decorate the function with `@torch.compile` so it's on by default?

File: examples/sphinx_gallery/local_padding.py (hunk 3 above)
Comment (akshaysubr) [reply examples/sphinx_gallery/local_padding.py:160]: This is cool! :)

File: tests/test_healpix_padding.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    lon = torch.from_numpy(grid.lon)
+    z = lon + 3 * lat
+    z = torch.arange(grid.shape[0])
+
+    z = z.reshape(1, 12, nside, nside).float()
+
+    with pad_backend(PaddingBackends.zephyr):
+        expected = pad(z, padding=nside)
+
+    with pad_backend(PaddingBackends.indexing):
```
Comment (akshaysubr) [inline tests/test_healpix_padding.py:56]: Would it also make sense to add the cuda backend to this test?

**Review by akshaysubr — APPROVED** [review]

File: earth2grid/healpix/_padding/pure_python.py (hunk 4 above)
Comment (nbren12) [reply earth2grid/healpix/_padding/pure_python.py:78]: confusing names, but it is correct.

File: earth2grid/healpix/nn.py (hunk 5 above)
Comment (nbren12) [reply earth2grid/healpix/nn.py:34]: hmm. I just moved  this code as is so will leave changes to future prs.

File: earth2grid/healpix/nn.py (hunk 6 above)
Comment (nbren12) [reply earth2grid/healpix/nn.py:38]: hmm. I just moved  this code as is so will leave changes to future prs.

File: tests/test_healpix_padding.py (hunk 8 above)
Comment (nbren12) [reply tests/test_healpix_padding.py:56]: that wouldn't be able to run on github actions, so I didn't bother. in the medium term, after testing this code, i would like to remove both the cuda and zephyr implementation. The regression tests should let us do that.
